pkg/cloudlogging/puller_test.go

```diff
@@ -98,7 +98,69 @@ func TestPull(t *testing.T) {
 				t.Errorf("Process(%+v) got result diff (-want, +got): %v", tc.name, diff)
 			}
 
-			if diff := cmp.Diff(tc.wantReq, tc.server.req, protocmp.Transform()); diff != "" {
+			if diff := cmp.Diff(tc.wantReq, tc.server.listReq, protocmp.Transform()); diff != "" {
+				t.Errorf("Process(%+v) got request diff (-want, +got): %v", tc.name, diff)
+			}
+		})
+	}
+}
+
+func TestSteamPull(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name          string
+		server        *fakeServer
+		filter        string
+		wantReq       *loggingpb.TailLogEntriesRequest
+		wantResult    []*loggingpb.LogEntry
+		wantErrSubstr string
+	}{
+		{
+			name:   "success",
+			filter: "test-filter",
+			server: &fakeServer{
+				tailResp: &loggingpb.TailLogEntriesResponse{
+					Entries: []*loggingpb.LogEntry{{LogName: "test"}},
+				},
+			},
+			wantResult: []*loggingpb.LogEntry{{LogName: "test"}},
+		},
+		{
+			name:   "failed_to_pull",
+			filter: "test-filter",
+			server: &fakeServer{
+				injectedErr: fmt.Errorf("injected error"),
+			},
+			wantErrSubstr: "injected error",
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			ctx := context.Background()
+			fakeClient := setupFakeClient(t, ctx, tc.server)
+			p := NewPuller(
+				ctx,
+				fakeClient,
+				testResource,
+				WithRetry(retry.WithMaxRetries(0, retry.NewFibonacci(500*time.Millisecond))),
+			)
+			ch := make(chan []*loggingpb.LogEntry, 1)
+			gotErr := p.SteamPull(ctx, tc.filter, 1, ch)
```

Comment:
Close the ch after it is done? close(ch)

Reply:
I think from the official, channel should be closed by sender instead of receiver, so I think it should be closed by `streampull()`

https://go.dev/tour/concurrency/4